The job **`aarch64-gnu-llvm-21-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/24302388691/job/70957928200) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/70957928200) [(plain)](https://github.com/rust-lang/rust/commit/27f6f66d41a1fefb7ac8af354b4c39e87e1a55bd/checks/70957928200/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Executing "/scripts/stage_2_test_set1.sh"
+ /scripts/stage_2_test_set1.sh
PR_CI_JOB set; skipping tidy
+ '[' 1 == 1 ']'
+ echo 'PR_CI_JOB set; skipping tidy'
+ SKIP_TIDY='--skip tidy'
+ ../x.py --stage 2 test --skip tidy --skip compiler --skip src
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://static.rust-lang.org/dist/2026-03-05/rustfmt-nightly-aarch64-unknown-linux-gnu.tar.xz
---

---- [codegen] tests/codegen-llvm/issues/issue-144329-niched-option-check.rs stdout ----
------rustc stdout------------------------------

------rustc stderr------------------------------
error[E0601]: `main` function not found in crate `issue_144329_niched_option_check`
##[error]  --> /checkout/tests/codegen-llvm/issues/issue-144329-niched-option-check.rs:42:2
   |
42 | }
   |  ^ consider adding a `main` function to `/checkout/tests/codegen-llvm/issues/issue-144329-niched-option-check.rs`

---
------------------------------------------

error: compilation failed!
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/codegen-llvm/issues/issue-144329-niched-option-check.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "-O" "-Cdebug-assertions=no" "-Zcodegen-source-order" "--emit" "llvm-ir" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/issues/issue-144329-niched-option-check/issue-144329-niched-option-check.ll" "-A" "internal_features" "-A" "incomplete_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Copt-level=3"
stdout: none
--- stderr -------------------------------
error[E0601]: `main` function not found in crate `issue_144329_niched_option_check`
##[error]  --> /checkout/tests/codegen-llvm/issues/issue-144329-niched-option-check.rs:42:2
   |
```

</details>

--- a/tests/codegen-llvm/issues/issue-144329-niched-option-check.rs
+++ b/tests/codegen-llvm/issues/issue-144329-niched-option-check.rs
@@ -1,0 +1,44 @@
+//! Ensure that redundant null checks on `&mut T` from `Option<(_, &mut T)>` are eliminated.
+
+//@ compile-flags: -Copt-level=3
+
+#![crate_type = "lib"]
+
+type T = [u64; 4];
+
+// CHECK-LABEL: @f(
+#[no_mangle]
+pub fn f(stack: &mut Stack, f: fn(&T)) -> bool {
+    // CHECK-NOT: icmp eq ptr
+    let Some((_a, b)) = stack.popn_top::<0>() else {
+        return false;
+    };
+    f(b);
+    true
+}
+
+pub struct Stack {
+    data: Vec<T>,
+}
+
+impl Stack {
+    #[inline(always)]
+    fn popn_top<const N: usize>(&mut self) -> Option<([T; N], &mut T)> {
+        if self.data.len() < N + 1 {
+            return None;
+        }
+        unsafe { Some((self.popn_unchecked(), self.top_unchecked())) }
+    }
+
+    unsafe fn popn_unchecked<const N: usize>(&mut self) -> [T; N] {
+        core::array::from_fn(|_| unsafe { self.pop_unchecked() })
+    }
+
+    unsafe fn pop_unchecked(&mut self) -> T {
+        self.data.pop().unwrap_unchecked()
+    }
+
+    unsafe fn top_unchecked(&mut self) -> &mut T {
+        self.data.last_mut().unwrap_unchecked()
+    }
+}

